go/consensus/cometbft/apps/roothash/messages_test.go

```diff
@@ -88,3 +91,360 @@ func TestChangeParameters(t *testing.T) {
+	require.EqualValues(blk3.Header.IORoot, roots[3].IORoot)
+
+	rfr2, err := state.RoundRoots(ctx, runtime.ID, 2)
+	require.NoError(err, "RoundRoots 2")
+	require.Nil(rfr2, "roots for round 2 should have been deleted in round 3")
+
+	// Switch context.
+	ctx.Close()
+	ctx = appState.NewContext(abciAPI.ContextEndBlock)
+
+	// Prepare & apply proposal for increasing the number of roots stored.
+	changeMaxPastRootsStored(require, app, ctx, state, 2)
+
+	// Switch context.
+	ctx.Close()
+	ctx = appState.NewContext(abciAPI.ContextBeginBlock)
+
+	// Round 4 (2 root pairs stored).
+	blk4 := advanceRuntimeState(require, ctx, blk, blk3, runtime, 4)
+	roots, err = state.PastRoundRoots(ctx, runtime.ID)
+	require.NoError(err, "PastRoundRoots")
+	require.EqualValues(2, len(roots))
+	require.EqualValues(len(roots), state.PastRoundRootsCount(ctx, runtime.ID))
+	require.EqualValues(blk3.Header.StateRoot, roots[3].StateRoot)
+	require.EqualValues(blk3.Header.IORoot, roots[3].IORoot)
+	require.EqualValues(blk4.Header.StateRoot, roots[4].StateRoot)
+	require.EqualValues(blk4.Header.IORoot, roots[4].IORoot)
+
+	// Round 5 (2 root pairs stored).
+	blk5 := advanceRuntimeState(require, ctx, blk, blk4, runtime, 5)
+	roots, err = state.PastRoundRoots(ctx, runtime.ID)
+	require.NoError(err, "PastRoundRoots")
+	require.EqualValues(2, len(roots))
+	require.EqualValues(len(roots), state.PastRoundRootsCount(ctx, runtime.ID))
+	require.EqualValues(blk4.Header.StateRoot, roots[4].StateRoot)
+	require.EqualValues(blk4.Header.IORoot, roots[4].IORoot)
+	require.EqualValues(blk5.Header.StateRoot, roots[5].StateRoot)
+	require.EqualValues(blk5.Header.IORoot, roots[5].IORoot)
+
+	rfr3, err := state.RoundRoots(ctx, runtime.ID, 3)
+	require.NoError(err, "RoundRoots 3")
+	require.Nil(rfr3, "roots for round 3 should have been deleted in round 5")
+
+	// Round 6 (2 root pairs stored).
+	blk6 := advanceRuntimeState(require, ctx, blk, blk5, runtime, 6)
+	roots, err = state.PastRoundRoots(ctx, runtime.ID)
+	require.NoError(err, "PastRoundRoots")
+	require.EqualValues(2, len(roots))
+	require.EqualValues(len(roots), state.PastRoundRootsCount(ctx, runtime.ID))
+	require.EqualValues(blk5.Header.StateRoot, roots[5].StateRoot)
+	require.EqualValues(blk5.Header.IORoot, roots[5].IORoot)
+	require.EqualValues(blk6.Header.StateRoot, roots[6].StateRoot)
+	require.EqualValues(blk6.Header.IORoot, roots[6].IORoot)
+
+	rfr4, err := state.RoundRoots(ctx, runtime.ID, 4)
+	require.NoError(err, "RoundRoots 4")
+	require.Nil(rfr4, "roots for round 4 should have been deleted in round 6")
+
+	// Switch context.
+	ctx.Close()
+	ctx = appState.NewContext(abciAPI.ContextEndBlock)
+
+	// Prepare & apply proposal for disabling storing the roots.
+	changeMaxPastRootsStored(require, app, ctx, state, 0)
+
+	// Switch context.
+	ctx.Close()
+	ctx = appState.NewContext(abciAPI.ContextBeginBlock)
+
+	// Round 7 (0 root pairs stored).
+	_ = advanceRuntimeState(require, ctx, blk, blk6, runtime, 7)
```

Comment:
Do we need to advance to test shrink?

Reply:
It's nicer this way and more closely models how the code will be used in real life, rather than hacking something together :)

Reply:
The problem that I saw here was that some roots should be removed immediately after `changeMaxPastRootsStored` is called, but in your test we also call `advanceRuntimeState` which calls `SetRuntimeState` which removes the oldest root. So I cannot be sure that `changeMaxPastRootsStored` removed the roots correctly.

Reply:
OK, I'll add an extra test.